phper/src/types.rs

```diff
@@ -10,162 +10,282 @@
 }
 
 impl TypeInfo {
-    pub const fn from_raw(t: u32) -> Self {
-        Self { t }
+    pub fn varadic(&mut self) -> &Self {
+        self.inner.type_mask |= _ZEND_IS_VARIADIC_BIT;
+
+        self
+    }
+
+    pub const fn from_type_mask(mask: u32) -> Self {
+        Self {
+            inner: ZendType {
+                ptr: std::ptr::null::<std::ffi::c_void>() as *mut std::ffi::c_void,
+                type_mask: mask,
+            },
+        }
     }
 
-    pub const fn into_raw(self) -> u32 {
-        self.t
+    pub fn from_zval(zval: &ZVal) -> Self {
+        let mask = unsafe { phper_z_type_info_p(zval.as_ptr()) };
+        if (mask & IS_OBJECT) != 0 {
+            let name_ptr = unsafe {
+                let obj = phper_z_obj_p(zval.as_ptr());
+                let ce = (*obj).ce;
+                phper_zend_object_release(obj);
+                (*ce).name as *mut std::ffi::c_void
+            };
+
+            Self {
+                inner: ZendType {
+                    ptr: name_ptr,
+                    type_mask: mask | _ZEND_TYPE_NAME_BIT,
+                },
+            }
+        } else {
+            Self::from_type_mask(mask)
+        }
+    }
+
+    pub const fn from_class_entry(ce: &zend_class_entry) -> Self {
+        let name_ptr = ce.name as *mut std::ffi::c_void;
+
+        Self {
+            inner: ZendType {
+                ptr: name_ptr,
+                type_mask: _ZEND_TYPE_NAME_BIT,
+            },
+        }
+    }
+
+    pub const fn from_raw(inner: ZendType) -> Self {
+        Self { inner }
+    }
+
+    pub const fn into_raw(self) -> ZendType {
+        self.inner
     }
 
     pub const fn undef() -> TypeInfo {
-        Self::from_raw(IS_UNDEF)
+        Self::from_type_mask(IS_UNDEF)
     }
 
     pub const fn null() -> TypeInfo {
-        Self::from_raw(IS_NULL)
+        Self::from_type_mask(IS_NULL)
     }
 
     pub const fn bool(b: bool) -> TypeInfo {
-        Self::from_raw(if b { IS_TRUE } else { IS_FALSE })
+        Self::from_type_mask(if b { IS_TRUE } else { IS_FALSE })
     }
 
     pub const fn long() -> TypeInfo {
-        Self::from_raw(IS_LONG)
+        Self::from_type_mask(IS_LONG)
     }
 
     pub const fn double() -> TypeInfo {
-        Self::from_raw(IS_DOUBLE)
+        Self::from_type_mask(IS_DOUBLE)
     }
 
     pub const fn string() -> TypeInfo {
-        Self::from_raw(IS_STRING)
+        Self::from_type_mask(IS_STRING)
     }
 
     pub const fn array() -> TypeInfo {
-        Self::from_raw(IS_ARRAY)
+        Self::from_type_mask(IS_ARRAY)
     }
 
     pub const fn array_ex() -> TypeInfo {
-        Self::from_raw(IS_ARRAY_EX)
+        Self::from_type_mask(IS_ARRAY_EX)
     }
 
     pub const fn object() -> TypeInfo {
-        Self::from_raw(IS_OBJECT)
+        Self::from_type_mask(IS_OBJECT)
     }
 
     pub const fn object_ex() -> TypeInfo {
-        Self::from_raw(IS_OBJECT_EX)
+        Self::from_type_mask(IS_OBJECT_EX)
+    }
+
+    pub const fn is_complex(self) -> bool {
```

Comment:
returns true if: type refers to a ce, union, or intersection.

Reply:
FWIW, these inline PR comments should probably be code comments :)